netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -851,54 +843,113 @@ public AsciiString scheme() {
    * subsequent handlers to assume the channel is active and ready to send.  Additionally, this a
    * {@link ProtocolNegotiationEvent}, with the connection addresses.
    */
-  static final class WaitUntilActiveHandler extends ChannelInboundHandlerAdapter {
-    private final ChannelHandler next;
-    private ProtocolNegotiationEvent pne;
+  static final class WaitUntilActiveHandler extends ProtocolNegotiationHandler {
 
-    public WaitUntilActiveHandler(ChannelHandler next) {
-      this.next = checkNotNull(next, "next");
-    }
+    boolean protocolNegotiationEventReceived;
 
-    @Override
-    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-      negotiationLogger(ctx).log(ChannelLogLevel.INFO, "WaitUntilActive started");
-      // This should be a noop, but just in case...
-      super.handlerAdded(ctx);
+    WaitUntilActiveHandler(ChannelHandler next) {
+      super(next);
     }
 
     @Override
     public void channelActive(ChannelHandlerContext ctx) throws Exception {
+      if (protocolNegotiationEventReceived) {
+        replaceOnActive(ctx);
+        fireProtocolNegotiationEvent(ctx);
+      }
       // Still propagate channelActive to the new handler.
       super.channelActive(ctx);
-      if (pne != null) {
+    }
+
+    @Override
+    protected void protocolNegotiationEventTriggered(ChannelHandlerContext ctx) {
+      protocolNegotiationEventReceived = true;
+      if (ctx.channel().isActive()) {
+        replaceOnActive(ctx);
         fireProtocolNegotiationEvent(ctx);
       }
     }
 
+    private void replaceOnActive(ChannelHandlerContext ctx) {
+      ProtocolNegotiationEvent existingPne = getProtocolNegotiationEvent();
+      Attributes attrs = existingPne.getAttributes().toBuilder()
+          .set(Grpc.TRANSPORT_ATTR_LOCAL_ADDR, ctx.channel().localAddress())
+          .set(Grpc.TRANSPORT_ATTR_REMOTE_ADDR, ctx.channel().remoteAddress())
+          // Later handlers are expected to overwrite this.
+          .set(GrpcAttributes.ATTR_SECURITY_LEVEL, SecurityLevel.NONE)
+          .build();
+      replaceProtocolNegotiationEvent(existingPne.withAttributes(attrs));
+    }
+  }
+
+  /**
+   * ProtocolNegotiationHandler is a convenience handler that makes it easy to follow the rules for
+   * protocol negotiation.  Handlers should strongly consider extending this handler.
+   */
+  static class ProtocolNegotiationHandler extends ChannelDuplexHandler {
```

Comment:
should this be abstract class to force override protocolNegotiationEventTriggered? 
(handlerAdded0 and other ForOverride ones could remain same)

Reply:
That's actually a good idea.  Let me change it.

Reply:
Hmm, well after trying it out, it doesn't seem so easy.  `protocolNegotiationEventTriggered` doesn't actually include the event (by design) so classes like ClientTlsHandler have to override it, even though they may not use it.